contracts/TokenHolder.sol

```diff
@@ -0,0 +1,511 @@
+     * @param _data the bytecode to be executed.
+     *         Use web3 getData method to construct _data.
+     * @param _v It's the recovery id.
+     * @param _r It's the output of an ECDSA signature.
+     * @param _s It's also the output of an ECDSA signature.
+     *
+     * @return Execution result is true/false.
+     */
+    function executeRule(
+        address _from,
+        address _to,
+        uint256 _nonce,
+        bytes _data,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    )
+        public
+        returns (bool executionResult_ /* success */)
+    {
+        EphemeralKeyData storage ephemeralKeyData = verifySignature(
+            _from,
+            _to,
+            _nonce,
+            _data,
+            _v,
+            _r,
+            _s
+        );
+
+        BrandedToken(brandedToken).approve(
+            tokenRules,
+            ephemeralKeyData.spendingLimit
+        );
+        executionResult_ = _to.call(_data);
+        emit RuleExecuted(_from, _to, _nonce, executionResult_);
+        BrandedToken(brandedToken).approve(
+            tokenRules,
+            0
+        );
+
+        return executionResult_;
+    }
+
+    /**
+     * @notice redeem after validating signature.
+     *
+     * @dev requestRedeem will call coGateway redeem.
+     *      coGateway redeem is a payable function so that bounty is deducted from msg.sender.
+     *      msg.sender is workers here.
+     *      It's TokenRule 0, CoGateway is fixed rule. e.g. Airdrop is TR.
+     *
+     * @param _from it enforces wallet to store TH address.
+     * @param _to the target contract the transaction will be executed upon.
+     *        e.g. CoGateway in case for CoGateway.redeem.
+     * @param _nonce incremental nonce.
+     * @param _data the bytecode to be executed.
+     *        It's the bytecode CoGateway redeem function.
+     * @param _v It's the recovery id.
+     * @param _r It's the output of an ECDSA signature.
+     * @param _s It's also the output of an ECDSA signature.
+     *
+     * @return Redeem execution result is success/failure.
+     */
+    function redeem(
+        address _from,
+        address _to,
+        uint256 _nonce,
+        bytes _data,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    )
+        public
+        payable
+        returns (bool executionResult_ /* success */)
+    {
+        EphemeralKeyData storage ephemeralKeyData = verifySignature(
+            _from,
+            _to,
+            _nonce,
+            _data,
+            _v,
+            _r,
+            _s
+        );
+        require(
+            _to == coGateway,
+            "to should be coGateway address."
+        );
+
+        BrandedToken(brandedToken).approve(
+            coGateway,
+            ephemeralKeyData.spendingLimit
+        );
+        // coGateway.redeem is a payable function.
+        executionResult_ = coGateway.call.value(msg.value)(_data);
```

Comment:
Should this confirm that `_data` is for a function that matches the redeem function signature on the co-gateway? Are there any concerns about a token holder contract calling any function on the co-gateway?

Relatedly, as this function does not check that `_data` is for redeem, if that will stay true (and perhaps even if it doesn't), this and `executeRule` are virtually identical and so they seem like prime candidates for refactoring.

Reply:
Let's discuss first point on call.

For second point we had discussion with @benjaminbollen. verifySignature internal function was refactored for the similar reasons. We concluded that having an explicit redeem function is good. There is an extra validation of coGateway address in redeem.